analyses/hello-clusters/01_perform-evaluate-clustering.Rmd

```diff
@@ -49,32 +62,443 @@ data_dir <- file.path(repository_base, "data", "current")
+
+We'll use the function `rOpenScPCA::calculate_purity()` to calculate the neighborhood purity.
+
+This function will return the inputted data frame with two additional columns:
+
+* `purity`: The neighborhood purity for the cell
+* `maximum_neighbor`: The cluster with the highest proportion of observations neighboring the cell
+
+
+```{r purity}
+# calculate the neighborhood purity for each cell
+purity_results <- calculate_purity(
+  pca_matrix,
+  cluster_results_df
+)
+
+# print result
+purity_results
+```
+
+### Cluster stability
+
+Another approach to exploring cluster quality is how stable the clusters themselves are using bootstrapping. 
+Given a set of original clusters, we can compare the bootstrapped cluster identities to original ones using the Adjusted Rand Index (ARI), which measures the similarity of two data clusterings.
+ARI ranges from -1 to 1, where:
+
+* A value of 1 indicates they are completely overlapping
+* A value of -1 indicates they are completely distinct
+* A value of 0 indicates a random relationship
+
+We expect that highly stable clusterings have ARI values closer to 1 across a set of bootstrap replicates.
+
+You can read more about the Adjusted Rand Index from the [_Orchestrating Single Cell Analysis with Bioconductor_ book](https://bioconductor.org/books/release/OSCA.advanced/clustering-redux.html#adjusted-rand-index).
+
+We'll use the function `rOpenScPCA::calculate_stability()` to calculate the cluster stability.
+By default, this function performs 20 bootstrap replicates, but this can be customized using the argument `replicates`.
+
+This function will return a data frame with columns `replicate`, `ari`, and additional columns for the clustering parameters used when calculating bootstrap clusters.
+
+```{r stability, warning=FALSE}
+# calculate the stability of clusters
+stability_results <- calculate_stability(
+  pca_matrix,
+  cluster_results_df
+)
+
+# print the result
+stability_results
+```
+
+#### Using non-default clustering parameters
+
+When calculating bootstrap clusters, `rOpenScPCA::calculate_stability()` uses `rOpenScPCA::calculate_clusters()` with default parameters. 
+If your original clusters were not calculated with these defaults, you should pass those customized values into this function as well to ensure a fair comparison between your original clusters and the bootstrap clusters.
+
+
+```{r stability custom parameters}
+# Calculate clusters with non-default parameters
+cluster_df_leiden <- calculate_clusters(
+  pca_matrix,
+  algorithm = "leiden",
+  resolution = 0.5,
+  nn = 15
+)
+
+# Now, pass in the same arguments customizing parameters here
+stability_results_leiden <- calculate_stability(
+  pca_matrix,
+  cluster_df_leiden,
+  algorithm = "leiden",
+  resolution = 0.5,
+  nn = 15
+)
+
+stability_results_leiden
+```
+
+
+
+##  Calculate QC metrics on existing clusters
```

Comment:
I would move this to after the object section and demonstrate these using the objects themselves.